### DISCUSSION TITLE
How to define a separate response_model for HTTP 400 errors?

### DISCUSSION THREAD

**Discussion by houmie** [fastapi/fastapi#2906] [Questions] [answered]
I'm forced to set all values in a ```response_model``` as Optional.

```
class ConnectOut(BaseModel):
    product_id: Optional[str]
    expires_at: Optional[datetime]
    detail: Optional[ErrorType]
```

If I don't do that the HTTP400 path below throws validation errors, because product_id and expires_at won't be provided in case of a 400 error.

```
@router_connect.post("/", status_code=200, response_model=ConnectOut)
async def connect(
    body: ConnectIn,
    response: Response,
):
    if account.is_banned:
        response.status_code = status.HTTP_400_BAD_REQUEST
        return {"detail": ErrorType.USER_IS_BANNED}
```

Is there a way to define a response_model for success and response_model for 400 Error messages?

Many Thanks,

**Comment by ec2ainun** [comment, marked as answer]
@houmie 
according to docs how to [Create Response][1]

> you still want to be able to filter and convert the data you return with a response_model

response_model created for filtering and converting returned value

by modifying your code, you could use [responses][2]
```py
@router_connect.post(
    "/", 
    summary="Get info product",
    responses={
        **DEFAULT_RESPONSE,
        200: {"model": yourmodel.RespData, "description": "Succeed Get data"},
        400: {"model": yourmodel.RespErr, "description": ErrorType.USER_IS_BANNED},
        },
    )
async def connect(
    body: ConnectIn,
):
    if account.is_banned:
        # refers to yourmodel.RespErr
        resp = {"detail": ErrorType.USER_IS_BANNED}
        return JSONResponse(
            content=resp, 
            status_code=400,
            headers=yourDefinedHeader,
            media_type=yourMimetype
        )
```


[1]: https://fastapi.tiangolo.com/advanced/response-change-status-code
[2]: https://fastapi.tiangolo.com/advanced/additional-responses

**Reply by houmie** [reply]
@ec2ainun 

Thank you so much for sharing this with me. I have found a similar solution with HTTPException, what do you think of this? 

```
class HTTPError(BaseModel):
    detail: int
```

```
@router_connect.post(
    "/",
    status_code=200,
    response_model=ConnectOut,
    responses={200: {"model": ConnectOut}, 400: {"model": HTTPError}},
)
async def connect(body: ConnectIn):
    if account.is_banned:
        raise HTTPException(status_code=400, detail=ErrorType.USER_IS_BANNED)
```

**Reply by ec2ainun** [reply]
@houmie 
according to docs how to [Handling Error][1]

>  it's a Python exception, you don't return it, you raise it.

it depends on the design and preference for each individual/company though 
if you prefer to make it like an error go for it

[1]: https://fastapi.tiangolo.com/tutorial/handling-errors
